### ISSUE TITLE
YOLOv8 Not Detecting Multiple Objects per Image

### ISSUE THREAD

**Issue by badrdaa2-afk** [ultralytics/ultralytics#22072]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### I trained a YOLOv8 model using a combined dataset:

Extracted the "bowl" class from COCO.

Merged it with a new dataset containing other classes.

Updated the dataset.yaml file and trained the model.

### the issue is :
The model detects objects, but when an image contains multiple classes ,  it detects only one. 
so its detect one classes per image

**Comment by UltralyticsAssistant** [comment]
👋 Hello @badrdaa2-afk, thanks for opening this issue and for using Ultralytics 🚀. This is an automated response — an Ultralytics engineer will review and assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for helpful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples and answers to common questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) so we can investigate. For your case of only one detection per image, please include:
-1–3 sample images with corresponding labels that reproduce the issue
-Exact commands used for train/val/predict
-Ultralytics package version and environment details (OS, Python, PyTorch, CUDA)
-Your dataset.yaml (or a minimal snippet showing class names and paths)
-Model used and checkpoint path
-Training logs and validation metrics
-Inference settings you used (e.g., confidence, IoU/NMS parameters, imgsz)
-Example prediction outputs showing the single detection behavior

If this is a custom training ❓ Question, please share dataset examples and training logs, and verify you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). 🙌

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
What's the code you used to train and make inference?

**Comment by badrdaa2-afk** [comment]
> What's the code you used to train and make inference?

from ultralytics import YOLO

model = YOLO("yolov8n.pt")
model.train(
    data="/kaggle/working/mynewdataset/dataset.yaml",  
    epochs=100,       
    imgsz=640,       
    batch=16,        
    name="tray_bowl_model", 
    patience=5
)

# Example inference
res = model.predict('/kaggle/input/image.jpg', conf=0.25, iou=0.5, max_det=100)
img = res[0].plot()

**Comment by Y-T-G** [comment]
Probably an issue with your dataset. Does your dataset have multiple classes in the same image?

**Comment by badrdaa2-afk** [comment]
> Probably an issue with your dataset. Does your dataset have multiple classes in the same image?

No , each image contain only one class

**Comment by Y-T-G** [comment]
Your model overfit to your dataset then to just predict one class per image.

**Comment by badrdaa2-afk** [comment]
what can i do to avoid overfitting ?

**Comment by Y-T-G** [comment]
You need to improve your dataset and have examples with multiple classes in the same image.

**Comment by badrdaa2-afk** [comment]
Thank you !

**Comment by glenn-jocher** [comment]
Glad it helped—along with adding multi-class images, boost co-occurrence via augmentation to reduce overfitting; for example:
```python
from ultralytics import YOLO
model = YOLO("yolov8n.pt")
model.train(data="/kaggle/working/mynewdataset/dataset.yaml",
            epochs=100, imgsz=640, batch=16, mosaic=1.0, mixup=0.2)
```
For concise guidance see the Ultralytics overview on overfitting and our YOLO augmentation reference: [Overfitting in Machine Learning (Ultralytics guide)](https://www.ultralytics.com/glossary/overfitting), [Ultralytics YOLO data augmentation techniques](https://docs.ultralytics.com/reference/data/augment/).
